### ISSUE TITLE
Datagrid hangs the browser when it re-renders several times with its dataSource property set to a different array value

### ISSUE THREAD

**Issue by yurigenin** [inovua/reactdatagrid#116]
- what edition are you using - enterprise
- version for `@inovua/reactdatagrid-enterprise` - 4.1.9

Relevant code or config

```javascript
import React from 'react'
import ReactDataGrid from '@inovua/reactdatagrid-enterprise'
import '@inovua/reactdatagrid-enterprise/index.css'

const columns = [
  { name: 'name', header: 'Name', minWidth: 50, defaultFlex: 2 },
  { name: 'age', header: 'Age', maxWidth: 1000, defaultFlex: 1 }
]

const gridStyle = { minHeight: 550 }

const dataSource = [
  { id: 1, name: 'John McQueen', age: 35 },
  { id: 2, name: 'Mary Stones', age: 25 },
  { id: 3, name: 'Robert Fil', age: 27 },
  { id: 4, name: 'Roger Robson', age: 81 },
  { id: 5, name: 'Billary Konwik', age: 18 },
  { id: 6, name: 'Bob Martin', age: 18 },
  { id: 7, name: 'Matthew Richardson', age: 54 },
  { id: 8, name: 'Ritchie Peterson', age: 54 },
  { id: 9, name: 'Bryan Martin', age: 40 },
  { id: 10, name: 'Mark Martin', age: 44 },
  { id: 11, name: 'Michelle Sebastian', age: 24 },
  { id: 12, name: 'Michelle Sullivan', age: 61 },
  { id: 13, name: 'Jordan Bike', age: 16 },
  { id: 14, name: 'Nelson Ford', age: 34 },
  { id: 15, name: 'Tim Cheap', age: 3 },
  { id: 16, name: 'Robert Carlson', age: 31 },
  { id: 17, name: 'Johny Perterson', age: 40 }
]


export default () => {
  const [count, setCount] = React.useState(2);
  
  React.useEffect(() => {
    if (count !== 0) {
        setCount(count - 1);
    }
  }, [count]);

  const currentDataSource = count !== 0 ? [] : dataSource;

  console.log('rendered with count: ', count, ' and dataSource: ', currentDataSource);
  
  return (
      <ReactDataGrid
         idProperty="id"
         columns={columns}
         dataSource={currentDataSource}
         style={gridStyle}
        />
   );
}
```

What you did:
This is a pretty vanilla code. On first two iterations the grid is passed an empty data source array and on the third iteration an array of 17 items. At this point the grid just hangs and to the point that you have to kill the page. 

What happened:
Basically, it goes into an infinite loop. I tried to debug it and it seems that it is stuck in the useDataSource hook and can never get out of it. (The code is very hard to understand so eventually I just bailed. It would be great if you guys went through it and commented it because it seems like it is one of the most important areas in terms of performance. But when you look at the call stack in this instance you will see the same code frame re-entered 6 or 7 times and it just keeps on executing it endlessly).

The application workaround fix would be not to pass a new empty array on every iteration but use a constant empty array. But this is not really a solution since this is something a lot of developers would do by convention and the grid has to handle such a scenario. I understand that passing a new reference every time affects performance but it should never hang  the browser, especially that it is only three renders is what it takes to hang it. I tried with a non-empty array and the result is the same.

About half of our grids experienced this behavior on upgrade as using an in-place empty array is very common before the data get fetched. But this would even happen if you just pass a new non-empty dataSource reference three times in a row which some applications may easily do.

As you see in the console panel, the application renders three times successfully:

![image](https://user-images.githubusercontent.com/11202960/115966292-1c5b9a80-a4ea-11eb-9fb6-afc135280d74.png)

But the component ends up in the infinite loop. As you see below, the same block executes over and over:

![image](https://user-images.githubusercontent.com/11202960/115966394-9ee45a00-a4ea-11eb-8f69-257a6f066e78.png)

**Comment by inovua-admin** [comment]
Should be fixed in `4.1.13`

**Comment by yurigenin** [comment]
Thank you!
